test/integration/examples_test.go

```diff
@@ -30,6 +30,16 @@ const examplesDIR = "../../examples"
 var httprouteExampleManifests = fmt.Sprintf("%s/gateway-httproute.yaml", examplesDIR)
 
 func TestHTTPRouteExample(t *testing.T) {
+	_, cleaner := setup(t)
+	defer func() {
+		if t.Failed() {
+			output, err := cleaner.DumpDiagnostics(ctx, t.Name())
+			t.Logf("%s failed, dumped diagnostics to %s", t.Name(), output)
+			assert.NoError(t, err)
+		}
+		assert.NoError(t, cleaner.Cleanup(ctx))
```

Comment:
non-blocking: This block is really making me pine for a `cleaner.CleanupWithDiagnostics(ctx, t)` method which checks `t.Failed()` itself and dumps the diagnostics if true.

Reply:
My original thinking here was that:
- Tests may want to do something before calling diagnostics, or add conditions beyond test failure. This is probably a non-issue, because they still control their `defer` contents, so they can do something else or wrap another if around the call.
- The other calls to cleaner methods (and KTF in general) do not take a `testing.T` and fence all logic related to that outside, so I didn't want to pass one and pierce that veil. Do you think it'd suffice to instead use a call like `cleaner.Whatever(ctx, t.Name(), t.Failed()` (the last argument would be `runDiagnostics bool` in the definition)? That preserves the separation while still providing a single-line invocation.